### PR TITLE
do not discover properties when iterating fixtures

### LINKED ISSUE #12446: pytest 8.2.0 calls `@property`s during unittest collection
Since 8.2.0 and changes to [unittest collecting](https://github.com/pytest-dev/pytest/commit/1a5e0eb71d2af0ad113ccd9ee596c7d724d7a4b6) pytests behaves now slightly differently. If there is  @property defined in unittest class it is called during tests collection. This leads to some absurd behaviour with each property being collected twice. Attached example takes 20s to collect instead of 0.01s

```
import unittest
from time import sleep

class TestClass(unittest.TestCase):

    @property
    def variable(self):
        print('My collection sleep')
        sleep(10)


class MyTest(TestClass):
    def test_simple(self):
        assert True
```

pip list output:

```
pip list
Package                   Version
------------------------- ---------
attrs                     23.2.0
cffi                      1.16.0
cryptography              42.0.8
exceptiongroup            1.2.1
iniconfig                 2.0.0
jsonschema                4.22.0
jsonschema-specifications 2023.12.1
packaging                 24.1
pip                       22.0.4
pluggy                    1.5.0
pycparser                 2.22
pyOpenSSL                 24.1.0
pytest                    8.2.2
referencing               0.35.1
rpds-py                   0.18.1
setuptools                58.1.0
tomli                     2.0.1
```

### PR DESCRIPTION
**Pull request by asottile**
resolves #12446

### CODE REVIEW

File: src/_pytest/fixtures.py [hunk 1]
```diff
@@ -1724,17 +1725,26 @@ def parsefactories(
         if holderobj in self._holderobjseen:
             return
 
+        # avoid accessing `@property` (and other descriptors) when iterating fixtures
```
Comment (nicoddemus) [inline src/_pytest/fixtures.py:1728]:
```suggestion
        # Avoid accessing `@property` (and other descriptors) when iterating fixtures.
```

File: src/_pytest/fixtures.py [hunk 2]
```diff
@@ -1724,17 +1725,26 @@ def parsefactories(
...
-            marker = getfixturemarker(obj)
+            # access below can raise. safe_getattr() ignores such exceptions.
+            obj_ub = safe_getattr(holderobj_tp, name, None)
+            marker = getfixturemarker(obj_ub)
             if not isinstance(marker, FixtureFunctionMarker):
                 # Magic globals  with __getattr__ might have got us a wrong
                 # fixture attribute.
                 continue
 
+            # ok we know it is a fixture -- now safe to look up on the _instance_
```
Comment (nicoddemus) [inline src/_pytest/fixtures.py:1745]:
```suggestion
            # OK we know it is a fixture -- now safe to look up on the _instance_.
```

File: testing/python/collect.py [hunk 3]
```diff
@@ -263,6 +263,41 @@ def prop(self):
         result = pytester.runpytest()
         assert result.ret == ExitCode.NO_TESTS_COLLECTED
 
+    def test_does_not_discover_properties(self, pytester: Pytester) -> None:
```
Comment (nicoddemus) [inline testing/python/collect.py:266]:
```suggestion
    def test_does_not_discover_properties(self, pytester: Pytester) -> None:
        """Regression test for #12446."""
```

File: testing/python/collect.py [hunk 4]
```diff
@@ -263,6 +263,41 @@ def prop(self):
...
+            class TestCase:
+                @property
+                def oops(self):
+                    raise SystemExit('do not call me!')
+            """
+        )
+        result = pytester.runpytest()
+        assert result.ret == ExitCode.NO_TESTS_COLLECTED
+
+    def test_does_not_discover_instance_descriptors(self, pytester: Pytester) -> None:
```
Comment (nicoddemus) [inline testing/python/collect.py:278]:
```suggestion
    def test_does_not_discover_instance_descriptors(self, pytester: Pytester) -> None:
        """Regression test for #12446."""    
```

**Review by nicoddemus — APPROVED** [review]
Looks great, thanks @asottile!

Left just some minor fixes/suggestions.

**Comment by RonnyPfannschmidt** [comment]
Nice one, I tried that in https://github.com/pytest-dev/pytest/pull/6898 before but it went stale

**Review by RonnyPfannschmidt — APPROVED** [review]

**Pull request merged**